File: src/main/java/com/tms/csp/ast/Not.java
package com.tms.csp.ast;

import com.google.common.collect.Iterators;
import com.tms.csp.argBuilder.ArgBuilder;
import com.tms.csp.fm.dnnf.products.Cube;
import com.tms.csp.util.varSets.VarSet;

import javax.annotation.Nonnull;
import java.util.Iterator;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;

public class Not extends Complex {

    private final PosComplexMultiVar pos;
    private boolean isNew;

    public Not(@Nonnull Exp pos, int expId) {
        super(pos.getSpace(), expId);
        checkNotNull(pos);
        checkArgument(pos.isPosComplex());
        this.pos = pos.asPosComplex();
        this.isNew = true;
    }


    @Override
    public Exp flip() {
        return pos;
    }

    @Override
    public boolean isPos() {
        return false;
    }

    @Override
    public Exp getPos() {
        return pos;
    }

    @Override
    public Exp getNeg() {
        return this;
    }

    @Override
    public boolean isNot() {
        return true;
    }

    @Override
    public boolean isComplex() {
        return pos.isComplex();
    }

    @Override
    public boolean anyVarOverlap(Exp exp) {
        return pos.anyVarOverlap(exp);
    }

    @Override
    public Exp pushNotsIn() {
        Exp ret =  pushNotsInInternal();
        //        System.err.println("Not.pushNotsIn: ");
//        System.err.println("  Before: " + this);
//        System.err.println("  After: " + ret);

        return ret;
    }

    private Exp pushNotsInInternal() {
        if (pos.isAnd()) {
            pos.asAnd().createHardFlip();
            return pos.asAnd().createHardFlip();
        } else if (pos.isOr()) {
            return pos.asOr().createHardFlip();
        } else {
            return this;
        }
    }

    @Override
    public boolean isNegComplex() {
        return isComplex();
    }

    @Override
    public PosOp getPosOp() {
        return pos.getPosOp();
    }

    @Override
    public Var getFirstVar() {
        return pos.getFirstVar();
    }


    @Override
    public boolean containsVarId(int varId) {
        return pos.containsVarId(varId);
    }

    @Override
    public boolean isOrContainsConstant() {
        return pos.isOrContainsConstant();
    }

    @Override
    public Op getOp() {
        return Op.Not;
    }

    @Override
    public Exp condition(Cube cube) {
        assert !isConstant();

        Exp sPos = pos.condition(cube);

        if (sPos == pos) {
                        return this;
        } else {
                        return sPos.flip();
        }

    }

    public Exp simplify() {
        throw new UnsupportedOperationException();
//        EvalContext ctx = null;
//        Exp simplify = pos.simplify(ctx);
//        return simplify.flip();
    }


    @Override
    public Exp condition(Lit lit) {

//        System.err.println("Not.condition(lit)");
        if (!containsVar(lit)) {
            return this;
        }


        Exp sPos = pos.condition(lit);

        if (sPos.isNot()) {
                        if (true) throw new IllegalStateException();
            return sPos.getArg();
        } else {
            return sPos.flip();
        }


    }

    @Override
    final public Exp getArg(int i) {
        if (i == 0) return pos;
        throw new IndexOutOfBoundsException();
    }

    @Override
    public int getArgCount() {
        return 1;
    }


    public Iterator<Exp> argIter() {
        Exp pp = pos;
        return Iterators.singletonIterator(pp);
    }

    @Override
    public VarSet getVars() {
        return pos.getVars();
    }

    public void serialize(Ser a) {
        if (pos.isConstantTrue()) {
            a.constantFalse();
        } else if (pos.isPosComplex()) {
            a.bang();
            pos.serialize(a);
        } else {
            System.err.println("Should never get here: Not.serialize");
            throw new IllegalStateException();
        }
    }

    public Exp hardFlip() {
        System.err.println("Not.hardFlip");
        PosComplexMultiVar pos = getPos().asPosComplex();

        ArgBuilder b = pos.flipArgs();
        Space space = getSpace();
        if (pos.isAnd()) {
            assert b.getOp().isAnd();
            return b.mk();
        }
        if (pos.isOr()) {
            assert b.getOp().isOr();
            return b.mk();
        }

        return this;
    }

    public Exp toNnf() {
        assert !isXorOrContainsXor();
        Exp pNnf = pos.toNnf(false);
        assert pNnf.isOr() || pNnf.isAnd();

        Op newOp;
        if (pos.isOr()) {
            newOp = Op.And;
        } else {
            newOp = Op.Or;
        }
        ArgBuilder b = new ArgBuilder(_space, newOp);
        for (Exp arg : pNnf.argIt()) {
            b.addExp(arg.flip());
        }

        return b.mk();
    }


    @Override
    public boolean isSat() {
        return toNnf().isSat();
    }

    @Override
    public boolean hasFlip() {
        return true;
    }

    @Override
    public boolean isNew() {
        return isNew;
    }


    public void notNew() {
        isNew = false;
    }

    public boolean canPushNotsIn() {
        return pos.isOr() || pos.isAnd();
    }

    public boolean samePos(Not that) {
        return pos.equals(that.pos);
    }

    public Exp toDnnf() {
        Exp nnf = toNnf();
        assert !nnf.isNot();
        return nnf.toDnnf();
    }

    public long satCountPL() {
        return toNnf().satCountPL();
    }

}

File: src/main/java/com/smartsoft/csp/ast/Not.kt
package com.smartsoft.csp.ast

import com.google.common.collect.Iterators
import com.smartsoft.csp.argBuilder.ArgBuilder
import com.smartsoft.csp.dnnf.products.Cube
import com.smartsoft.csp.varSets.VarSet

class Not(pos: Exp, expId: Int) : Complex(pos.space, expId) {

    override val pos: PosComplexMultiVar

    override var isNew: Boolean = false
        private set

    override val isPos: Boolean
        get() = false

    override val arg: Exp get() = pos

    override val neg: Exp get() = this

    override val isNot: Boolean get() = true

    override val isComplex: Boolean get() = true

    override val isNegComplex: Boolean get() = true

    override val posOp: PLConstants.PosOp get() = pos.posOp

    override val firstVar: Var get() = pos.firstVar

    override val op: Op get() = Op.Not

    override val argCount: Int get() = 1

    override val vars: VarSet get() = pos.vars


    override val isSat: Boolean get() = toNnf().isSat

    init {
        assert(pos is PosComplexMultiVar)
        this.pos = pos as PosComplexMultiVar
        this.isNew = true
    }

//    fun flip(): Exp {
//        return pos
//    }

    override val flip: Exp get() = pos

    override fun anyVarOverlap(exp: Exp): Boolean {
        return pos.anyVarOverlap(exp)
    }

    override fun pushNotsIn(): Exp {
        val ret = pushNotsInInternal()
        //        System.err.println("Not.pushNotsIn: ");
        //        System.err.println("  Before: " + this);
        //        System.err.println("  After: " + ret);

        return ret
    }

    private fun pushNotsInInternal(): Exp {
        if (pos.isAnd) {
            pos.asAnd.createHardFlip()
            return pos.asAnd.createHardFlip()
        } else return if (pos.isOr) {
            pos.asOr.createHardFlip()
        } else {
            this
        }
    }


    override fun containsVarId(varId: Int): Boolean {
        return pos.containsVarId(varId)
    }

    override fun condition(ctx: Cube): Exp {
        assert(!isConstant)

        val sPos = pos.condition(ctx)

        return if (sPos === pos) {
            this
        } else {
            sPos.flip
        }

    }

    fun simplify(): Exp {
        throw UnsupportedOperationException()
        //        EvalContext ctx = null;
        //        Exp simplify = pos.simplify(ctx);
        //        return simplify.flip;
    }


    override fun condition(lit: Lit): Exp {

        //        System.err.println("Not.condition(lit)");
        if (!containsVar(lit)) {
            return this
        }


        val sPos = pos.condition(lit)

        if (sPos.isNot) {
            if (true) throw IllegalStateException()
            return sPos.arg
        } else {
            return sPos.flip
        }


    }

    override fun getArg(index: Int): Exp {
        if (index == 0) return pos
        throw IndexOutOfBoundsException()
    }


    fun argIter(): Iterator<Exp> {
        val pp = pos
        return Iterators.singletonIterator<Exp>(pp)
    }

    override fun serialize(a: Ser) {
        if (pos.isConstantTrue) {
            a.constantFalse()
        } else if (pos.isPosComplex) {
            a.bang()
            pos.serialize(a)
        } else {
            System.err.println("Should never get here: Not.serialize")
            throw IllegalStateException()
        }
    }

    fun hardFlip(): Exp {
        System.err.println("Not.hardFlip")

        val b = pos.flipArgs()
        if (pos.isAnd) {
            assert(b.op.isAnd)
            return b.mk()
        }
        if (pos.isOr) {
            assert(b.op.isOr)
            return b.mk()
        }

        return this
    }

    fun toNnf(): Exp {
        assert(!isXorOrContainsXor)
        val pNnf = pos.toNnf(false)
        assert(pNnf.isOr || pNnf.isAnd)

        val newOp: Op
        if (pos.isOr) {
            newOp = Op.And
        } else {
            newOp = Op.Or
        }
        val b = ArgBuilder(_space, newOp)
        for (arg in pNnf.argIt) {
            b.addExp(arg.flip)
        }

        return b.mk()
    }

    fun hasFlip(): Boolean {
        return true
    }


    override fun notNew() {
        isNew = false
    }

    fun canPushNotsIn(): Boolean {
        return pos.isOr || pos.isAnd
    }

    fun samePos(that: Not): Boolean {
        return pos == that.pos
    }

    override fun toDnnf(): Exp {
        val nnf = toNnf()
        assert(!nnf.isNot)
        return nnf.toDnnf()
    }

    override val satCountPL: Long get() {
        if(true) throw IllegalStateException()
        return toNnf().satCountPL
    }

    override val argIter: Iterator<Exp> get() = Iterators.singletonIterator(pos)

}
